src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/AbstractImportCompletionCacheServiceFactory.cs

```diff
@@ -0,0 +1,57 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+#nullable enable
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+
+namespace Microsoft.CodeAnalysis.Completion.Providers.ImportCompletion
+{
+    internal abstract class AbstractImportCompletionCacheServiceFactory<TProjectCacheEntry, TMetadataCacheEntry> : IWorkspaceServiceFactory
```

Comment:
can you explain what the generics are here for?  when/how do we supply different instantiations of this?  thanks!

Reply:
For type completion, they are the same type, since project and PE share the same cache entry implementation. For extension method completion, we don't have cache for PE, having separate generic types make it clearer (IMO)

Reply:
ok.  thanks.  #resolved.